### DISCUSSION TITLE
🗺️ fetcher, submit, and Form Changes

### DISCUSSION THREAD

**Discussion by ryanflorence** [remix-run/remix#7698] [Official RFCs]
> [!IMPORTANT]  
> This proposal has changed significantly since originally posted, some comments might not make sense

This proposal has three parts to it:

- Adding `<Form navigate={false} />` to create a fetcher instead of a navigation
- Adding `submit(data, { navigate: false })` to create a fetcher instead of a navigation
- Changing the internals of fetcher cleanup
- Introducing `useFetcher({ key })`, `<Form fetcherKey>` and `submit(d, { fetcherKey })`

## Background

Today fetcher cleanup is coupled to the React lifecycle. When a component with a `useFetcher` unmounts and the fetcher is pending, the request is cancelled and the fetcher is removed from internal router state. This decision was made to prevent fetchers from being orphaned and causing memory leaks, and the UI lifecycle seemed like a good place to do it.

However, sometimes you want the request to persist beyond the lifecycle of the component, especially with "out of place" optimistic UI and components that open and close like dialogs and menu buttons. The current design makes these kinds of UI more difficult than they should be to build.

Apps need a way to simply toss out a fetcher and not care about the result in-place, but still have access to in `useFetchers` for optimistic UI and even be able to "pick it back up" in some other part of the UI.

For example, here is an implementation of **Optimistic Add** today that we think is overly cumbersome and leaks implementation details of fetcher lifecycles.

First is a list that reads loader data and renders all three things: the form, the pending items, and the items from the database.

```tsx
function List() {
  // real items from the DB
  let { items } = useLoaderData();

  // manage some pending states
  let [submissions, setSubmissions] = useState<FormData[]>([]);

  return (
    <>
      <form
        onSubmit={(e) => {
          event.preventDefault();
          // push submission in to cause a new PendingItem to render
          setSubmissions([
            ...pendingSubmissions,
            new FormData(e.target),
          ]);
        }}
      >
        <input type="text" name="name" />
        <button type="submit">Save Item</button>
      </form>

      {/* Render the pending submissions */}
      {submissions.map((formData, index) => (
        <PendingItem
          key={index}
          formData={formData}
          onComplete={() => {
            // remove the completed pending item
            setPendingColumns(
              pendingSubmissions.filter((f) => f !== formData),
            );
          }}
        />
      ))}

      {/* Render the real items*/}
      {items.map((item) => (
        <Item key={item.id} name={item.name} />
      ))}
    </>
  );
}
```

The `PendingItem` component only exists to keep the fetcher alive until the request is complete.

```tsx
function PendingItem({ formData, onComplete }) {
  let fetcher = useFetcher();
  let name = formData.get("name");

  useEffect(() => {
    fetcher.submit(formData, { method: "post" });
  }, []);

  useEffect(() => {
    if (fetcher.state === "idle" && fetcher.data) {
      onComplete();
    }
  }, [fetcher]);

  return <Item pending name={name} />;
}
```

And finally the `Item` component isn't relevant, we'll just assume that might change what it renders based on being an optimistic version or not.

This extra work is only needed because the fetcher lifecycle is coupled to the component lifecycle. There are also some bugs for the observant reader that always creep in when you synchronize state.

### `<Form navigate={bool}>` and `submit(data, { navigate: bool })`

Today `<Form>` always navigates and `<fetcher.Form>` creates a fetcher and tracks its state. This proposal adds a `navigate` prop to `<Form>` that allows it to create a fetcher instead of navigating.

```tsx
<Form navigate={false} method="post">
  <input name="title" />
</Form>
```

Now instead of causing a navigation, an internal fetcher will be created, accessible from `useFetchers`.

```tsx
function SomeComp() {
  let [fetcher] = useFetchers();
  fetchers && fetcher.formData.has("title"); // true
}
```

This allows forms to simply kick off a fetcher and not care about the result, decoupling the fetcher from the lifecycle so that optimistic UIs can render until they are filled in with loader data after revalidation, or conditional UI like dialogs can simply close without cancelling the request and subsequent revalidation.

On the imperative end of the spectrum, `submit` is simply the imperative version of `<Form>`. It works the same here: use `navigate: false` and a fetcher will be created instead of a navigation.

### Fetcher Cleanup changes

If `<Form navigate={false}>` can kick of fetchers that live beyond the lifecycle of the component, we need a new way to clean them up. This proposal changes the internals of fetcher cleanup to decouple it from the React lifecycle.

Today `fetcher.data` is managed internally by the router _outside_ of React (which is why the React lifecycle for cleanup made sense).

This change will move the "complete" state to React, allowing the router to simply track fetchers while they are pending, giving us a new point of garbage collection decoupled from rendering.

This might be a "breaking bug" (😅) for `useFetchers`. In some cases, `useFetchers` returns fetchers that are "idle" and have `fetcher.data` available on them even though it was never intended to do so. Even the docs explain it differently than the behavior.

Most use cases we've seen and authored check some sort of pending information like `fetcher.formData` so we don't expect it to affect very many apps. It might, so we're still considering putting this all behind a future flag, but we'd rather not.

Even if it does lead to undesired behavior, `useFetchers` is ephemeral, so it's unlikely your app will "break break", just some ephemeral states you used to have won't be triggered anymore. We think you'll like the amount of code you can delete by moving to the API updates in this proposal.

## `useFetcher({ key })`

Lastly, this proposal adds a new `key` API for fetchers that allows you to pick up a fetcher in a different part of the tree from where it was created, or even share it across the tree.

```tsx
let fetcher = useFetcher({ key: 123 });
```

Additionally, you can use keys with `Form` and `submit`:

```tsx
<Form navigate={false} fetcherKey="123" />;
submit(formData, { navigate: false, fetcherKey: "123" });
```

Now an optimistic version of a record can pick up the fetcher that created it, or two distant parts of a tree can share a fetcher.

The fetcher will also have a new prop on `fetcher.key` which can make it easier to filter out the fetchers you want instead of using formData.

## Examples

**Optimistic Add to List**

We can revisit the optimistic add code and see how simple it gets:

```tsx
function List() {
  let { items } = useLoaderData();

  // get the optimistic items from the fetchers
  let submissions = useFetchers()
    .filter((fetcher) => fetcher.formData?.get("intent") === "add-item")
    .map((fetcher) => Object.fromEntries(fetcher.formData));

  return (
    <>
      {/* kick off a fetcher instead of navigating */}
      <Form navigate={false} method="post">
        <input type="text" name="name" />
        <button type="submit" name="intent" value="add-item">
          Save Item
        </button>
      </Form>

      {/* Render everything together, optimistic and real */}
      {submissions.concat(items).map((item, index) => (
        <Item
          key={item.id || index}
          pending={Boolean(item.id)}
          name={item.name}
        />
      ))}
    </>
  );
}
```

Form will create new fetchers, they'll be rendered the duration of their request/revalidation cycle, and then will be filled in by loader data when the action is complete.

For some _serious React_ you could create client IDs and use them as the element `key` prop and then they could be fully interactive, as long as the backend understands it.

**Optimistic Move**

Imagine a trello-like interface where cards can be dragged from one column to another. The card that is _dragged_ does not initiate the fetch, but rather the card being _dropped_ does. The `onDrop` even can get the ID of the dragged card and simply kick of a fetch that that id and key.

```tsx
function Card({ id }) {
  let submit = useSubmit();

  function handleDrop(event) {
    let { cardId } = JSON.parse(
      event.dataTransfer.getData("application/move-card"),
    );
    submit(
      { cardId: data },
      {
        navigate: false,
        // use a key so it can be picked up by the optimistic card
        key: `move-card:${cardId}`,
      },
    );
  }

  return <div onDrop={handleDrop} />;
}
```

Now the optimistic card in the new column can pick up the fetcher to render pending states or even use it for other operations.

```tsx
function Card({ id }) {
  // pick up the fetcher from the other column
  let fetcher = useFetcher({ key: `move-card:${id}` });
  let submit = useSubmit();

  function handleDrop() {
    // ...
  }

  return (
    <div
      // use it for pending UI, even if it's already optimistic
      className={fetcher.state === "idle" ? "" : "pending"}
      draggable
      onDrop={handleDrop}
      onDragStart={handleDragStart}
    />
  );
}
```

**Dialogs and Menus**

Dialogs and menus can simply be a form that closes on submission and the fetch won't be cancelled.

```tsx
function SomeRoute() {
  let [edit, setEdit] = useState(false);
  return (
    <div>
      <h1>Some page</h1>
      {edit ? <MyModal onSave={() => setEdit(false)} /> : null}
    </div>
  );
}

function MyModal({ onSave }) {
  return (
    <Modal>
      <Form navigate={false} onSubmit={onSave}>
        <label>
          <input type="checkbox" name="watermelon" />{" "}
          <span>I would like my watermelon pre-sliced</span>
        </label>
        <button type="submit">Save</button>
      </Form>
    </Modal>
  );
}
```

**Comment by ShafSpecs** [comment]
Love this RFC. +1 from me.

The main pro for me personally is the removal of the need for extra states and effects in your app to track fetcher state and data. Asides from that, I see no issue with this RFC since it doesn't deduct anything from original implementation

**Comment by ryanflorence** [comment]
Should note that `useNavigation` is basically a singleton that already has these capabilities. It doesn't matter which `<Form>` started the mutation, `useNavigation().formData` is available everywhere, so new components can pick up the pending states and unmounting components don't cancel the fetch.

**Reply by bronifty** [reply]
That's magical. I was going to say it looks like you're inserting a cache key somewhere, but you're passing the formData around as a global instead. Very cool.

**Comment by kentcdodds** [comment]
Yep. I'm on board! Thanks for solving this with such a simple solution

**Comment by zachdtaylor** [comment]
I like it! This'll make a few things I'm doing a lot simpler.

**Comment by rossipedia** [comment]
I was kinda "oh yeah I can see this being useful"... until this

> The persist option can also take a key:
> ```tsx
> let fetcher = useFetcher({ persist: "abc" });
> ```
> With this key, any other calls to useFetcher with the same key will share the same underlying fetcher, sharing the state and lifecycle.

took me over the edge into full blown convert. Yes. Please.

**Reply by brophdawg11** [reply]
I chatted with Ryan and just we're probably going to split `persist` out into `useFetcher({ key?: string, persist?: boolean })` so you have the full set of capabilities.  `key` is for sharing, `persist` is for clean up behavior.  Otherwise it was ambiguous when `persist:string` fetchers would clean up.

**Comment by rphlmr** [comment]
What would happen for `fetcher.load`?

**Reply by sergiodxa** [reply]
Nothing in the RFC suggest it will removed, it should keep working as usual.

**Comment by JClackett** [comment]
Yes please! all my fetchers that are in collapsable menu's will now just work without any workarounds 🙌🏾

**Comment by devstojko** [comment]
Perfect! This will also solve issue with form in dialog. Now you can unmount dialog instead of hiding it while fetcher is doing it's job.

@ryanflorence after this we need promise / lifecycle for fetchers and that's it. I can get rid of ugly useEffects

**Reply by ryanflorence** [reply]
If you had access to the promise, you would now need useEffect and useState to do anything useful with it--unless you want bugs in your app by messing with promises in event handlers. We don't expose the promises so that we can do cancellation correctly and avoid race conditions and other bugs that come with it, all things you would need to do yourself with a promise.

**Reply by JClackett** [reply]
Even just an onComplete callback to the useFetcher would eliminate essentially all my useEffects, does this have to with promises?

**Reply by devstojko** [reply]
@ryanflorence Yes you are right, we don't want to deal with that. 
The issue that we are facing in our app is when you have 
- dependent actions in dialog
- get options on demand before opening dialog

Recently we had to implement like:

Open dialog (modal) 

You have 3 select options.

1. When you select first useFetcher to load options for second one => pass result to second select options => select second select => useFetcher to call another loader to get data for third select.

2. There are also situation where we have multiple selects in dialogs and I want to load (4 fetchers and get data) once user click button to open modal instead of passing them as props and load them initially in loader. (because we don't know if user is going to click button and consume data).

So this is why I am looking for "proper" solution for this kind of job. Using useEffect to track status for fetcher is tricky sometimes and devs in my team often includes "suggested" props in dep array and you get bugs. 

Also there are situations when you have case above (get response from first dialog) and when it's done open another one with returned data...

I just want an "easy" solution when you useFetcher to know when it finished job  without using useEffect.

Problem that you are going to solve now with this persist change:
 - we have to define all fetchers in modals in order to pass them to the forms so we can track fetcher status and when it finish job - show toaster notification, close modal (Unmount from dom) instead of just hiding it 
 
 - sometimes you want to close modal right after you submit fetcher (atm you can just "hide" it) instead of unmounting because it will cancel fetcher

**Reply by devstojko** [reply]
What if useFetcher has callback function where it can emit changes through callback value (same useEffect structure response) and then inline  in callback you can use same condition that you would use in use effect to get what you want and need.

**Reply by ryanflorence** [reply]
@JClackett If a fetcher is inflight, and the user navigates away, unmounting the component, you have an opportunity to do cleanup in `useEffect`.

If we provided `onComplete` we also need a cleanup mechanism. In many ways it's just recreating useEffect.

```tsx
let fetcher = useFetcher();

useEffect(() => {
  if (fetcher.data?.ok) {
    // this is `onComplete`
    return () => {
      // with cleanup when you unmount
    }
  }
}, [fetcher]);
```

Also, Remix might cancel a fetcher if you submit/load it twice in a row, which is another time you need cleanup.

`useEffect` properly handles locally scoped variables when the effect requires other variables from your component (like some local state + state of the fetcher).

Non-fetcher state changes, cleanup, and the component lifecycle all need to be considered together. Adding `onSuccess`, `onComplete`, etc. is just working against the grain in React and would definitely need to synchronization and stale value bugs in your components.

**Comment by otaviobarros** [comment]
Perfect, exactly what I needed, no more workarounds and state hoisting 🙌

**Comment by sergiodxa** [comment]
This will enable a nice pattern here a fetcher could be defined globally with a Symbol key:

```ts
const key = Symbol()
export function useMyFetcher() {
  return useFetcher({ persist: key })
}
```

Now `useMyFetcher` could have a unique persisted key that nothing else can use, this can be useful for @kentcdodds' [Full Stack Components pattern](https://www.epicweb.dev/full-stack-components).

```ts
// app/routes/api.something.ts
const key = Symbol()

export async function loader() {
  return json(...)
}

export function Something() {
  let fetcher = useFetcher({ persist: key })
  return ...
}
```
And with this you can import Something which already knows how to fetcher the loader and it will always be a single instance.

If it needs to be rendered in a list, we could also use Symbols by setting it in a state instead of a UUID.

```ts
function Something() {
  let [key] = useState(() => Symbol());
  let fetcher = useFetcher({ persist: key });

  return ...
}
```
And now each instance can have a unique key

**Reply by kentcdodds** [reply]
With that in mind, it might be useful to be able to initialize the fetcher with data.

**Comment by frontsideair** [comment]
In a side project of mine, I have a custom `Form` wrapper that takes an additional `handler` prop, which is basically like this:

```tsx
function Form({ handler, ...props }) {
  const fetcher = useFetcher();
  
  useEffect(() => {
    // call handler optimistically with `fetcher.formData` or `fetcher.data`
  }, [handler, fetcher])
  
  return handler ? <Form {...props} /> : <fetcher.Form {...props} />
```

I take a look at this component from time to time and don't like it a lot, but it provides some kind of symmetry and I can add optimistic stuff by adding a single handler. `<Form navigate={false} />` part of the proposal reminds me of this need and may allow me to delete this hack.

**Comment by sergiodxa** [comment]
With Form supporting navigate={false} and fetcherKey, why would I need useFetcher?

Also how should I access the Form data if it used a fetcher? I imagine I can't use useActionData anymore, same for useNavigation. Will I need to grab it from useFetchers?

**Reply by ryanflorence** [reply]
> With Form supporting navigate={false} and fetcherKey, why would I need useFetcher?

`useFetcher` is when you need to read the state. If it's all local and unmounts don't matter, you just keep using it the way you always have.

`navigate={false}` is to send off a fetcher from somewhere in the UI that you don't care about the state. Some other spot, usually optimistic UI, might care about the state. If it does, you bring in fetcherKey to the originator and continue to `useFetcher({ key })` to "pick it up". If the optimistic UI doesn't need the fetcher state, you don't need `useFetcher`.

**Reply by ryanflorence** [reply]
> Also how should I access the Form data if it used a fetcher?

It's still on `useFetcher`. So if it's all local, nothing changes.

If a `<Form navigate={false}>` kicks off a fetcher, you'd have access to formData in `useFetchers` and well as any `useFetcher({ key })` that picks it up.

`useActionData` and `useNavigation` aren't involved at all with a `<Form navigate={false}>`.

**Comment by brophdawg11** [comment]
This is being implemented in https://github.com/remix-run/react-router/pull/10949 and https://github.com/remix-run/remix/pull/7704

**Reply by brophdawg11** [reply]
The above PR has been superseded by:
* https://github.com/remix-run/react-router/pull/10960
* https://github.com/remix-run/react-router/pull/10961
* https://github.com/remix-run/react-router/pull/10962

**Comment by ryanflorence** [comment]
I just worked with an experimental version of this and it's :chefs_kiss:

There were some use cases that were way too difficult that I'd been wanting to solve and this nails it.

https://twitter.com/ryanflorence/status/1715440901378720013

**Comment by bluefire2121** [comment]
I like that this would get rid of fetchers and forms with stale data, but I think this is a big enough change that it should be a feature flag.  What would be the need for `<fetcher.Form>` after this change? The way I understand it, `<fetcher.Form>` = fetcher and `<Form navigate={false}>` = fetcher. Is that correct?

**Reply by sergiodxa** [reply]
I asked something similar here https://github.com/remix-run/remix/discussions/7698#discussioncomment-7331869. While `<form navigate={false}>` will work as a fetcher you will need to find the data or state in `useFetchers` so using `<fetcher.Form>` is going to be useful for when you need the data or state in your UI.

**Reply by ryanflorence** [reply]
Yep. fetcher.Form gives you that state "in place" and those use cases don't go anywhere and no behaviors change.

This proposal is about when you need that state "out of place" and don't want the originating "in olace" UI to cancel it if it unmounts.

**Comment by sergiodxa** [comment]
This will also enable another thing, reset a fetcher.

```ts
let [key, setKey] = useState(() => Symbol())
let fetcher = useFetcher({ key })

function resetFetcher() {
  setKey(Symbol())
}
```

Now if you call `resetFetcher` a new symbol will be created which will be considered a different key and will make useFetcher get a new instance.

**Reply by thosmos** [reply]
This is still a workaround for a missing `fetcher.reset()`.  It's reasonable to want to tie the `key` to a data object ID, like if you're editing a user record with id `123` then you'd want to be able to use that as the key.  Using this workaround means losing that just to trick the fetcher into resetting its `.data` field.  I'm surprised that `.reset()` is such a resisted feature, when it tracks directly to the underlying html form functionality that you actually are blocked from by using a fetcher.Form with any reactive UI attached to the `.data` field.

Here's a more detailed rant about the missing `.reset()` on a closed PR thread: https://github.com/remix-run/remix/pull/3551#issuecomment-1398035053 and another comment here: https://github.com/remix-run/remix/discussions/2749#discussioncomment-4732465

**Comment by brophdawg11** [comment]
This was released in [React Router 6.18.0](https://github.com/remix-run/react-router/releases/tag/react-router%406.18.0) and [Remix 2.2.0](https://github.com/remix-run/remix/releases/tag/remix%402.2.0)

**Reply by kentcdodds** [reply]
Just gotta say these improvements are killer 🤘 Great work team!

**Comment by Shadowfaxenator** [comment]
`<Form navigate={false} />` how this works if JS is disabled? Won't it work? Or it'll just post data with navigation?

**Reply by brophdawg11** [reply]
Correct - it's the same as if you used `<fetcher.Form>` so it will post to your `action` as a document navigation

**Comment by rodolphoasb** [comment]
Was `fetcher.key` implemented?

Quoting the original RFC: "The fetcher will also have a new prop on `fetcher.key` which can make it easier to filter out the fetchers you want instead of using formData."

 I'm trying to use that to create an optimistic mutation with:

```
  if (fetcher.formData?.has("value")) {
    value = String(fetcher.formData.get("value"));
  }
```

![image](https://github.com/remix-run/remix/assets/8127858/d81270cf-8fb9-4d4e-867f-a27c74be6e28)

**Reply by brophdawg11** [reply]
Fetchers returned from `useFetchers` have a `key` property - this is the intended use-case for "filtering" as per the original RFC.

When using `useFetcher( key: 'whatever' })` - there is no need to filter since you already have the specific `whatever` fetcher, so key is not exposed there.  The key you pass to `useFetcher` effectively does this filtering/finding internally.

**Comment by heivo** [comment]
Is there any way to still manually cancel a request with this? Previously it was possible to just unmount the component that renders the fetcher.

A use case might be a large file upload that can be aborted by the user.

**Reply by brophdawg11** [reply]
Hm, right now there might not be a super ergonomic way to do that.  Off the top of my head - if you were to issue a separate request from the same fetcher key it would abort the in-flight one.  You could use a `clientAction`/`clientLoader` then to "intercept" that second request and just no-op without going to the server.

Something like: 

```js
export async function action({ request }) {
  // handle large file upload
}

export async function clientAction({ request, serverAction }) {
   // If we see an abort query param, then we don't actually want to submit anything here
   if (new URL(request.url).searchParams .has('abort')) {
     return;
  }
  return await serverAction()
}

function Component() {
  let fetcher = useFetcher({ key: 'something' })l

  // To submit the file  
  fetcher.submit(data, { method: 'post', encType: 'multipart/form-data' });

  // To abort an ongoing upload
  fetcher.submit(null, { method: 'post', encType: 'multipart/form-data', action: '.?abort' });
}
```

It could be better to use a `fetcher.load` and a `clientLoader` too to avoid triggering other revalidations

**Reply by heivo** [reply]
That's an interesting idea, I was not aware that a client action cancels the previous request when it doesn't call the server action. I'll try that, thank you very much.
